qlib/VscDebugAdapter.qm

```diff
@@ -0,0 +1,1248 @@
+    the rights to use, copy, modify, merge, publish, distribute, sublicense,
+    and/or sell copies of the Software, and to permit persons to whom the
+    Software is furnished to do so, subject to the following conditions:
+
+    The above copyright notice and this permission notice shall be included in
+    all copies or substantial portions of the Software.
+
+    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+    FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+    DEALINGS IN THE SOFTWARE.
+*/
+
+# minimum required Qore version
+%requires qore >= 0.8.13
+%requires(reexport) DebugUtil
+%requires DebugCmdLine
+%requires Util
+%requires json
+
+%require-types
+%enable-all-warnings
+%new-style
+%no-debugging
+
+module VscDebugAdapter {
+    version = "0.1";
+    desc = "Implements Visual Studio Code Debug Adapter protocol";
+    author = "Tomas Mandys <tomas.mandys@qoretechnologies.com>";
+    url = "http://qore.org";
+    license = "MIT";
+}
+
+/** @mainpage VscDebugAdapter Module
+
+    @tableofcontents
+
+    @section vscdebugadapterintro Introduction to the VscDebugAdapter Module
+
+    The %VscDebugAdapter module provides common functionality for implementing
+    Visual Studio Adapter https://microsoft.github.io/debug-adapter-protocol/.
+
+    Only subset of features is supported in Visual Studio Code IDE and in this module.
+
+    Module limitations:
+    - breakpoint server side hitcount not supported
+    - breakpoint logpoints
+    - VSC event is not raised when breakpoint is changed, e.g. via another debugger client
+    - conditional breakpoints, e.g. for particular thread
+    - structured variables, e.g. classes/objects
+    - for functional breakpoint the function must be provided with arguments types even a user variant does not exist
+    - values for top-level local variables are not provided as thread vars for each thread
+    - cannot provide program exit value
+
+    Visual Studio Code known issues:
+    - evaluate "watch" request is not executed every step so watch value might be out of date
+    - no module support in IDE
+    - display own data type and custom data editor
+    - exception info is asked but the value is not displayed
+    - some capabilities are not implemented
+
+    @subsection vda_v0_1
+
+    - the initial version of the VscDebugAdapter module
+*/
+
+#! the VscDebugAdapter namespace contains all the definitions in the VscDebugAdapter module
+public namespace VscDebugAdapter {
+
+    #! class that can be used for logging
+    public class VscDebugLogger inherits DebugLogger {
```

Comment:
Why `Logger` is not used? Seems like new Logger class was introduced in `DebugUtil.qm` with constant like `DUV_DEBUG` etc

Reply:
This is legacy of Qore 0.8.x where we had no Logger. If you want get rid of DebugLogger which is logical requirement then you need rewrite all qdbg-* scripts and debugger stuff in Qorus. It is beyond this issue not related to vscode support at all.

Reply:
issue #3423 opened for this